### PR TITLE
sparse: 0.5.0 -> 0.6.3 and fixes

### PR DESCRIPTION
**Pull request by ethorsoe**
Added sqlite and gtk+3 that are now wanted by sparse.

Sparse expects to find system includes at /usr/include redirect this
to glibc.dev.

Sparse uses gcc -print-file-name= extensively to find gcc includes, due
to split in nixos this direct to gcc-lib, which does not include the headers
so redirect to GCC_BASE as defined in derivation. There might be a better way
to do this, but I did not immediately find one.

Define PREFIX as Make parameter, as old sed expression was broken.

Slightly irrelevant ref.
https://bugzilla.redhat.com/show_bug.cgi?id=1114755

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by jonringer — COMMENTED** [review]
This is what I came up with:

```
-{ callPackage, fetchurl, lib, stdenv, gcc-unwrapped, glibc, gtk3, pkg-config, libxml2, llvm, perl, sqlite }:
+{ callPackage, fetchurl, lib, stdenv, gtk3, pkg-config, libxml2, llvm, perl, sqlite }:

-stdenv.mkDerivation rec {
+let
+  GCC_BASE = "${stdenv.cc.cc}/lib/gcc/${stdenv.hostPlatform.uname.processor}-unknown-linux-gnu/${stdenv.cc.cc.version}";
+in stdenv.mkDerivation rec {
   pname = "sparse";
   version = "0.6.3";

@@ -9,20 +11,20 @@ stdenv.mkDerivation rec {
     sha256 = "16d8c4dhipjzjf8z4z7pix1pdpqydz0v4r7i345f5s09hjnxpxnl";
   };

-  GCC_BASE = "${gcc-unwrapped}/lib/gcc/${stdenv.hostPlatform.uname.processor}-unknown-linux-gnu/${gcc-unwrapped.version}";
   preConfigure = ''
-    sed -i 's|"/usr/include"|"${glibc.dev}/include"|' pre-process.c
+    sed -i 's|"/usr/include"|"${stdenv.cc.libc.dev}/include"|' pre-process.c
     sed -i 's|qx(\$ccom -print-file-name=)|"${GCC_BASE}"|' cgcc
     makeFlags+=" PREFIX=$out"
   '';

   nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ gtk3.dev libxml2 llvm perl sqlite.dev ];
+  buildInputs = [ gtk3 libxml2 llvm perl sqlite ];
   doCheck = true;
   buildFlags = "GCC_BASE:=${GCC_BASE}";

   passthru.tests = {
     simple-execution = callPackage ./tests.nix { };
+    inherit GCC_BASE;
   };

   meta = {
   ```

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 134638` at ae7645de run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>sparse</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

llvm is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/analysis/sparse/default.nix:21:3:
```
   |
21 |   buildInputs = [ gtk3 libxml2 llvm perl sqlite ];
   |   ^
```
</li>
  </ul>
</details>

**Comment by jkarlson** [comment]
rebased and added myself to maintainers, if that helps

File: pkgs/development/tools/analysis/sparse/tests.nix [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  set -eu
+  ${sparse}/bin/cgcc ${src} > output 2>&1 || ret=$?
+  if [[ -z $(<output) ]]; then
+    mv output $out
+  else
+    echo "Test build returned $ret"
+    cat output
+    exit 1
+  fi
+''
```
Comment (Artturin) [inline pkgs/development/tools/analysis/sparse/tests.nix:16]:
```suggestion
{ runCommand, gcc, sparse, writeText }:
let
  src = writeText "CODE.c" ''
    #include <stdio.h>
    #include <stddef.h>
    #include <stdlib.h>
    
    int main(int argc, char *argv[]) {
      return EXIT_SUCCESS;
    }
  '';
in
  runCommand "${sparse.pname}-tests" { buildInputs = [ gcc sparse ]; meta.timeout = 3; }
''
  set -eu
  ${sparse}/bin/cgcc ${src} > output 2>&1 || ret=$?
  if [[ -z $(<output) ]]; then
    mv output $out
  else
    echo "Test build returned $ret"
    cat output
    exit 1
  fi
''

```

+ remove trivial.c

**Comment by Artturin** [comment]
@cole-h ofborg internal error on this and some other PRs

is there a page where the error can be seen?

**Comment by cole-h** [comment]
Feel free to kick off a fresh evaluation whenever you see the label (just please don't remove it). The logs are not publicly visible because it may leak secrets. These days, the label basically means "GitHub responded with an incomplete API response which caused ofborg to choke".

@ofborg eval

**Pull request merged**